crates/sui-network/src/state_sync/mod.rs

```diff
@@ -1046,6 +1068,76 @@ where
     Ok(())
 }
 
+async fn sync_checkpoint_contents_from_archive<S>(
+    network: anemo::Network,
+    archive_readers: ArchiveReaderBalancer,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    metrics: Metrics,
+) where
+    S: WriteStore + Clone + Send + Sync + 'static,
+    <S as ReadStore>::Error: std::error::Error,
+{
+    loop {
+        let peers: Vec<_> = peer_heights
+            .read()
+            .unwrap()
+            .peers_on_same_chain()
+            // Filter out any peers who we aren't connected with.
+            .filter_map(|(peer_id, info)| network.peer(*peer_id).map(|peer| (peer, *info)))
+            .collect();
+        let lowest_checkpoint_on_peers = peers
+            .iter()
+            .map(|(_p, state_sync_info)| state_sync_info.lowest)
+            .min();
+        let highest_synced = store
+            .get_highest_synced_checkpoint()
+            .expect("store operation should not fail")
+            .sequence_number;
+        let sync_from_archive = if let Some(lowest_checkpoint_on_peers) = lowest_checkpoint_on_peers
+        {
+            highest_synced < lowest_checkpoint_on_peers
```

Comment:
I'm thinking what is the desired behavior here - when currently connected of peers do not have the checkpoint contents, should we encourage the peer to look for other peers, or simply turn to archives? The latter sounds appealing because we know it's faster. But it seems to put too much dependency on the archives. Wdyt?

Reply:
Oh interesting, I was not aware it is possible to make a peer look for other peers (some kind of gossip?) too. Fallback to archive seems reasonable to do, although if a node does not have access to archive for some reason it might make sense to have it also seek out other nodes until it finds a one which can help it catch up. Let's discuss this option offline too.